flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -2424,11 +2438,55 @@ struct GlobalOpConversion : public FIROpConversion<fir::GlobalOp> {
     auto g = rewriter.create<mlir::LLVM::GlobalOp>(
         loc, tyAttr, isConst, linkage, global.sym_name(), initAttr);
     auto &gr = g.getInitializerRegion();
+    if (global.getType().isa<fir::SequenceType>() &&
+        !global.getType().dyn_cast<fir::SequenceType>().hasUnknownShape() &&
+        global.hasInitializationBody()) {
+      if (auto singleInsertOp = getSingleOpOfType<fir::InsertOnRangeOp>(
+              global.region().front())) {
+        auto globalType = global.getType().dyn_cast<fir::SequenceType>();
+        // Check if the initialization is on the full range.
+        if (isFullRange(singleInsertOp.coor(), globalType)) {
+          rewriter.inlineRegionBefore(global.region(), gr, gr.end());
+          auto insertOp = getSingleOpOfType<fir::InsertOnRangeOp>(gr.front());
+          auto op = insertOp.val().getDefiningOp();
+          auto constant = dyn_cast<mlir::ConstantOp>(op);
+          if (!constant) {
+            auto convertOp = dyn_cast<fir::ConvertOp>(op);
+            constant = 
+                cast<mlir::ConstantOp>(convertOp.value().getDefiningOp());
+          }
+          mlir::Type vecType =
+              mlir::VectorType::get(globalType.getShape(), constant.getType());
+          auto denseAttr = mlir::DenseElementsAttr::get(
+              vecType.cast<ShapedType>(), constant.getValue());
+          rewriter.setInsertionPointAfter(insertOp);
+          rewriter.replaceOpWithNewOp<mlir::ConstantOp>(insertOp, tyAttr,
+                                                        denseAttr);
+          rewriter.eraseOp(global);
+          g.typeAttr(mlir::TypeAttr::get(tyAttr));
+          return success();
+        }
+      }
+    }
     rewriter.inlineRegionBefore(global.region(), gr, gr.end());
     rewriter.eraseOp(global);
     return success();
   }
 
+  bool isFullRange(mlir::ArrayAttr indexes, fir::SequenceType seqTy) const {
+    assert(indexes.size() >= 2 && indexes.size() % 2 == 0);
```

Comment:
Isn't this enforced in the verifier for `insert_on_range`?

Reply:
Looks like it's not at the moment.

Reply:
OK. That's a bug.

Reply:
I can move this assertion in the verify function of the insert_on_range op in this PR if that makes sense.

Reply:
Yeah. I think we should verify that `insert_on_range` is properly constructed as early as possible. The number of range indices has to be even. They have to be twice the rank of the `seq` argument's type, in fact.